File: backend/app/schemas/advisory.py
"""원클릭 분석·매매 (Advisory) Pydantic v2 스키마.

`/api/analysis/runs` 라우터의 요청·응답 스키마 정의.
금액/수량 응답 값은 마스킹 토글이 프론트에서 처리하므로 평문 Decimal로 노출한다.
"""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.config import settings
from app.models.trading import (
    AnalysisItemAction,
    AnalysisItemDecision,
    AnalysisItemSource,
    AnalysisRun,
    AnalysisRunItem,
    AnalysisRunStatus,
    TradingMode,
)
from app.services.crypto_service import (
    decrypt_decimal,
    decrypt_decimal_optional,
)


# ──────────────────────────────────────────────
# 요청 스키마
# ──────────────────────────────────────────────

class RunCreateRequest(BaseModel):
    """POST /api/analysis/runs — run 생성 본문."""

    account_id: UUID
    mode: TradingMode
    budget_krw: Decimal = Field(..., gt=Decimal("0"))
    candidate_pool_options: dict[str, Any] | None = None

    @field_validator("budget_krw")
    @classmethod
    def _budget_within_limit(cls, v: Decimal) -> Decimal:
        # 1회 run 예산 상한 — 운영 사고 방지용 안전장치.
        # 절대 상한은 ADVISORY_MAX_RUN_BUDGET_KRW (env) 로 설정.
        max_budget = settings.advisory_max_run_budget_krw
        if v > max_budget:
            raise ValueError(
                f"1회 run 예산은 {max_budget}원을 초과할 수 없습니다."
            )
        return v


class DecisionUpdate(BaseModel):
    """종목별 approve/reject 결정 1건."""

    item_id: UUID
    # 클라이언트가 보내는 결정. APPROVED/REJECTED 만 허용.
    decision: AnalysisItemDecision

    @field_validator("decision")
    @classmethod
    def _only_approve_or_reject(cls, v: AnalysisItemDecision) -> AnalysisItemDecision:
        if v not in (AnalysisItemDecision.APPROVED, AnalysisItemDecision.REJECTED):
            raise ValueError("decision 은 approved/rejected 만 허용합니다.")
        return v


class DecisionsRequest(BaseModel):
    """POST /api/analysis/runs/{id}/decisions — 일괄 결정 본문."""

    decisions: list[DecisionUpdate] = Field(..., min_length=1)


class ExecuteRequest(BaseModel):
    """POST /api/analysis/runs/{id}/execute — 발주 본문.

    live 모드는 totp_code 필수. paper 모드는 무시.
    """

    totp_code: str | None = None


# ──────────────────────────────────────────────
# 응답 스키마
# ──────────────────────────────────────────────

class AnalysisRunItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=False)

    id: UUID
    ticker: str
    ticker_name: str
    source: AnalysisItemSource
    action: AnalysisItemAction
    confidence: int
    reason: str | None
    ref_price: Decimal | None
    suggested_qty: Decimal | None
    decision: AnalysisItemDecision
    blocked_reason: str | None
    order_id: UUID | None


class AnalysisRunResponse(BaseModel):
    model_config = ConfigDict(from_attributes=False)

    id: UUID
    account_id: UUID
    mode: TradingMode
    status: AnalysisRunStatus
    budget_krw: Decimal
    candidate_pool_options: dict[str, Any]
    started_at: datetime
    completed_at: datetime | None
    expires_at: datetime | None
    summary_json: dict[str, Any] | None
    error_message: str | None
    items: list[AnalysisRunItemResponse]
    expired: bool


class AnalysisRunSummaryResponse(BaseModel):
    """이력 목록용 요약 — items 미포함."""

    model_config = ConfigDict(from_attributes=False)

    id: UUID
    account_id: UUID
    mode: TradingMode
    status: AnalysisRunStatus
    started_at: datetime
    completed_at: datetime | None
    expires_at: datetime | None
    item_count: int


class AnalysisRunListResponse(BaseModel):
    items: list[AnalysisRunSummaryResponse]
    total: int
    limit: int
    offset: int


class ExecuteItemResult(BaseModel):
    """발주 결과 1건."""

    item_id: UUID
    ticker: str
    action: AnalysisItemAction
    success: bool
    order_id: UUID | None = None
    error_message: str | None = None


class ExecuteResponse(BaseModel):
    run_id: UUID
    executed: int
    failed: int
    results: list[ExecuteItemResult]


# ──────────────────────────────────────────────
# 모델 → 스키마 변환 헬퍼
# ──────────────────────────────────────────────

def item_to_response(item: AnalysisRunItem) -> AnalysisRunItemResponse:
    return AnalysisRunItemResponse(
        id=item.id,
        ticker=item.ticker,
        ticker_name=item.ticker_name,
        source=item.source,
        action=item.action,
        confidence=item.confidence,
        reason=item.reason,
        ref_price=decrypt_decimal_optional(item.ref_price),
        suggested_qty=decrypt_decimal_optional(item.suggested_qty),
        decision=item.decision,
        blocked_reason=item.blocked_reason,
        order_id=item.order_id,
    )


def run_to_response(
    run: AnalysisRun,
    items: list[AnalysisRunItem],
    *,
    now: datetime,
) -> AnalysisRunResponse:
    expired = (
        run.expires_at is not None
        and run.status == AnalysisRunStatus.READY
        and run.expires_at < now
    )
    return AnalysisRunResponse(
        id=run.id,
        account_id=run.account_id,
        mode=run.mode,
        status=run.status,
        budget_krw=decrypt_decimal(run.budget_krw),
        candidate_pool_options=run.candidate_pool_options or {},
        started_at=run.started_at,
        completed_at=run.completed_at,
        expires_at=run.expires_at,
        summary_json=run.summary_json,
        error_message=run.error_message,
        items=[item_to_response(it) for it in items],
        expired=expired,
    )


def run_to_summary(run: AnalysisRun, item_count: int) -> AnalysisRunSummaryResponse:
    return AnalysisRunSummaryResponse(
        id=run.id,
        account_id=run.account_id,
        mode=run.mode,
        status=run.status,
        started_at=run.started_at,
        completed_at=run.completed_at,
        expires_at=run.expires_at,
        item_count=item_count,
    )

File: frontend/src/types/advisory.ts
// 원클릭 분석·매매 (Advisory) 타입.
// 백엔드 `app/schemas/advisory.py` 와 1:1 매칭.

import type { TradingMode } from "@/types/trading";

export type AnalysisRunStatus =
  | "pending"
  | "analyzing"
  | "ready"
  | "executing"
  | "done"
  | "failed";

export type AnalysisItemSource = "holding" | "auto_pick";
export type AnalysisItemAction = "buy" | "sell" | "hold";
export type AnalysisItemDecision =
  | "pending"
  | "approved"
  | "skipped"
  | "rejected";

export interface CandidatePoolOptions {
  top_n?: number;
  market?: "KOSPI" | "KOSDAQ" | "ALL";
  min_volume_value?: number;
  blacklist?: string[];
}

// ── API 응답 (snake_case) ──

export interface AnalysisRunItemApi {
  id: string;
  ticker: string;
  ticker_name: string;
  source: AnalysisItemSource;
  action: AnalysisItemAction;
  confidence: number;
  reason: string | null;
  ref_price: number | string | null;
  suggested_qty: number | string | null;
  decision: AnalysisItemDecision;
  blocked_reason: string | null;
  order_id: string | null;
}

export interface AnalysisRunApi {
  id: string;
  account_id: string;
  mode: TradingMode;
  status: AnalysisRunStatus;
  budget_krw: number | string;
  candidate_pool_options: CandidatePoolOptions;
  started_at: string;
  completed_at: string | null;
  expires_at: string | null;
  summary_json: Record<string, unknown> | null;
  error_message: string | null;
  items: AnalysisRunItemApi[];
  expired: boolean;
}

export interface AnalysisRunSummaryApi {
  id: string;
  account_id: string;
  mode: TradingMode;
  status: AnalysisRunStatus;
  started_at: string;
  completed_at: string | null;
  expires_at: string | null;
  item_count: number;
}

export interface AnalysisRunListApi {
  items: AnalysisRunSummaryApi[];
  total: number;
  limit: number;
  offset: number;
}

export interface ExecuteItemResultApi {
  item_id: string;
  ticker: string;
  action: AnalysisItemAction;
  success: boolean;
  order_id: string | null;
  error_message: string | null;
}

export interface ExecuteResponseApi {
  run_id: string;
  executed: number;
  failed: number;
  results: ExecuteItemResultApi[];
}

// ── 프론트 타입 (camelCase) ──

export interface AnalysisRunItem {
  id: string;
  ticker: string;
  tickerName: string;
  source: AnalysisItemSource;
  action: AnalysisItemAction;
  confidence: number;
  reason: string | null;
  refPrice: number | null;
  suggestedQty: number | null;
  decision: AnalysisItemDecision;
  blockedReason: string | null;
  orderId: string | null;
}

export interface AnalysisRun {
  id: string;
  accountId: string;
  mode: TradingMode;
  status: AnalysisRunStatus;
  budgetKrw: number;
  candidatePoolOptions: CandidatePoolOptions;
  startedAt: string;
  completedAt: string | null;
  expiresAt: string | null;
  summaryJson: Record<string, unknown> | null;
  errorMessage: string | null;
  items: AnalysisRunItem[];
  expired: boolean;
}

export interface AnalysisRunSummary {
  id: string;
  accountId: string;
  mode: TradingMode;
  status: AnalysisRunStatus;
  startedAt: string;
  completedAt: string | null;
  expiresAt: string | null;
  itemCount: number;
}

export interface AnalysisRunList {
  items: AnalysisRunSummary[];
  total: number;
  limit: number;
  offset: number;
}

export interface ExecuteItemResult {
  itemId: string;
  ticker: string;
  action: AnalysisItemAction;
  success: boolean;
  orderId: string | null;
  errorMessage: string | null;
}

export interface ExecuteResponse {
  runId: string;
  executed: number;
  failed: number;
  results: ExecuteItemResult[];
}

// ── 요청 타입 ──

export interface RunCreateRequest {
  account_id: string;
  mode: TradingMode;
  budget_krw: number;
  candidate_pool_options?: CandidatePoolOptions;
}

export interface DecisionUpdateRequest {
  item_id: string;
  decision: "approved" | "rejected";
}

export interface DecisionsRequest {
  decisions: DecisionUpdateRequest[];
}

export interface ExecuteRequest {
  totp_code?: string;
}

// ── 변환 함수 ──

function toNumberOrNull(v: number | string | null | undefined): number | null {
  if (v === null || v === undefined) return null;
  const n = typeof v === "number" ? v : Number(v);
  return Number.isFinite(n) ? n : null;
}

export function toAnalysisRunItem(api: AnalysisRunItemApi): AnalysisRunItem {
  return {
    id: api.id,
    ticker: api.ticker,
    tickerName: api.ticker_name,
    source: api.source,
    action: api.action,
    confidence: api.confidence,
    reason: api.reason,
    refPrice: toNumberOrNull(api.ref_price),
    suggestedQty: toNumberOrNull(api.suggested_qty),
    decision: api.decision,
    blockedReason: api.blocked_reason,
    orderId: api.order_id,
  };
}

export function toAnalysisRun(api: AnalysisRunApi): AnalysisRun {
  return {
    id: api.id,
    accountId: api.account_id,
    mode: api.mode,
    status: api.status,
    budgetKrw: Number(api.budget_krw),
    candidatePoolOptions: api.candidate_pool_options ?? {},
    startedAt: api.started_at,
    completedAt: api.completed_at,
    expiresAt: api.expires_at,
    summaryJson: api.summary_json,
    errorMessage: api.error_message,
    items: (api.items ?? []).map(toAnalysisRunItem),
    expired: api.expired,
  };
}

export function toAnalysisRunSummary(
  api: AnalysisRunSummaryApi,
): AnalysisRunSummary {
  return {
    id: api.id,
    accountId: api.account_id,
    mode: api.mode,
    status: api.status,
    startedAt: api.started_at,
    completedAt: api.completed_at,
    expiresAt: api.expires_at,
    itemCount: api.item_count,
  };
}

export function toAnalysisRunList(api: AnalysisRunListApi): AnalysisRunList {
  return {
    items: (api.items ?? []).map(toAnalysisRunSummary),
    total: api.total,
    limit: api.limit,
    offset: api.offset,
  };
}

export function toExecuteResponse(api: ExecuteResponseApi): ExecuteResponse {
  return {
    runId: api.run_id,
    executed: api.executed,
    failed: api.failed,
    results: (api.results ?? []).map((r) => ({
      itemId: r.item_id,
      ticker: r.ticker,
      action: r.action,
      success: r.success,
      orderId: r.order_id,
      errorMessage: r.error_message,
    })),
  };
}

// ── 라벨/판정 헬퍼 ──

export const runStatusLabels: Record<AnalysisRunStatus, string> = {
  pending: "준비 중",
  analyzing: "분석 중",
  ready: "결과 준비",
  executing: "발주 중",
  done: "완료",
  failed: "실패",
};

export const itemActionLabels: Record<AnalysisItemAction, string> = {
  buy: "매수",
  sell: "매도",
  hold: "관망",
};

export const itemDecisionLabels: Record<AnalysisItemDecision, string> = {
  pending: "대기",
  approved: "승인",
  // "차단" 은 blocked_reason 의미와 충돌해 UI 혼동을 일으킨다 — 자동 마킹된 항목은 "스킵".
  skipped: "스킵",
  rejected: "거절",
};

export const itemSourceLabels: Record<AnalysisItemSource, string> = {
  holding: "보유",
  auto_pick: "자동선정",
};

export const RUN_IN_PROGRESS_STATUSES: AnalysisRunStatus[] = [
  "pending",
  "analyzing",
  "executing",
];

export function isRunInProgress(status: AnalysisRunStatus): boolean {
  return RUN_IN_PROGRESS_STATUSES.includes(status);
}
